src/iter/collect/mod.rs

```diff
@@ -57,60 +54,70 @@ where
     right.truncate(0);
 
     let len = pi.len();
-    let mut left = Collect::new(left, len);
-    let mut right = Collect::new(right, len);
-
-    unzip_indexed(pi, left.as_consumer(), right.as_consumer());
-
-    left.complete();
-    right.complete();
+    Collect::new(left, len).with_consumer(|left_consumer| {
+        let mut left_result = None;
+        Collect::new(right, len).with_consumer(|right_consumer| {
+            let (left_r, right_r) = unzip_indexed(pi, left_consumer, right_consumer);
+            left_result = Some(left_r);
+            right_r
+        });
+        left_result.unwrap()
+    });
```

Comment:
The left and right parts cannot be mixed up. First, because they correspond to type parameters `A` and `B`, but even with that difference removed, the borrow checker will stop the user from mixing up left and right results.